cpp/src/groupby/hash/groupby_kernels.cuh

```diff
@@ -63,88 +63,59 @@ namespace hash {
  * @tparam Map The type of the hash map
- * @param map Hash map object to insert key,value pairs into.
- * @param num_keys The number of rows in input keys table
- * @param input_values The table whose rows will be aggregated in the values of
- * the hash map
- * @param output_values Table that stores the results of aggregating rows of
- * `input_values`.
- * @param aggs The set of aggregation operations to perform accross the columns
- * of the `input_values` rows
- * @param row_bitmask Bitmask where bit `i` indicates the presence of a null
- * value in row `i` of input keys. Only used if `skip_rows_with_nulls` is `true`
  */
 template <bool skip_rows_with_nulls, typename Map>
-__global__ void compute_single_pass_aggs(
-    Map map, size_type num_keys, table_device_view input_values,
-    mutable_table_device_view output_values, aggregation::Kind* aggs,
-    bitmask_type const* const __restrict__ row_bitmask)
-{  
-  size_type i = threadIdx.x + blockIdx.x * blockDim.x;
+struct compute_single_pass_aggs {
+  Map map;
+  size_type num_keys;
+  table_device_view input_values;
+  mutable_table_device_view output_values; 
+  aggregation::Kind const* __restrict__ aggs;
+  bitmask_type const*  __restrict__ row_bitmask;
 
-  while (i < num_keys) {
-    if (skip_rows_with_nulls and not cudf::bit_is_set(row_bitmask, i)) {
-      i += blockDim.x * gridDim.x;
-      continue;
-    }
+  /**
+   * @brief Construct a new compute_single_pass_aggs functor object
+   * 
+   * @param map Hash map object to insert key,value pairs into.
+   * @param num_keys The number of rows in input keys table
+   * @param input_values The table whose rows will be aggregated in the values 
+   * of the hash map
+   * @param output_values Table that stores the results of aggregating rows of
+   * `input_values`.
+   * @param aggs The set of aggregation operations to perform accross the 
+   * columns of the `input_values` rows
+   * @param row_bitmask Bitmask where bit `i` indicates the presence of a null
+   * value in row `i` of input keys. Only used if `skip_rows_with_nulls` is `true`
+   */
+  compute_single_pass_aggs(
+    Map map,
+    size_type num_keys,
+    table_device_view input_values,
+    mutable_table_device_view output_values, 
+    aggregation::Kind const* aggs,
+    bitmask_type const* row_bitmask)
+  : map(map),
+    num_keys(num_keys),
+    input_values(input_values),
+    output_values(output_values),
+    aggs(aggs),
+    row_bitmask(row_bitmask)
+  {}
 
-    auto result = map.insert(thrust::make_pair(i, i));
+  __device__ void operator()(size_type i) {
+    if (not skip_rows_with_nulls or
+        cudf::bit_is_set(row_bitmask, i))
+    {
+      auto result = map.insert(thrust::make_pair(i, i));
 
-    experimental::detail::aggregate_row<true, true>(
-      output_values, result.first->second, input_values, i, aggs);
-    i += blockDim.x * gridDim.x;
+      experimental::detail::aggregate_row<true, true>(
```

Comment:
Shouldn't the template args here be based on whether or not the keys/values actually contain nulls?

Reply:
I felt like it added code complexity and likely compile time for little benefit. The template argument can be set to false only when none of the columns are nullable. And even then, it only saves a null pointer check.

Mainly, I feel the template arguments `bool target_has_nulls, bool source_has_nulls` are unnecessary. It doesn't indicate if source or target actually has nulls, just that it may. `update_target_element` has to be able to handle either case anyway. 

See `aggregate_row` is called on a table. It should ideally select the value of this variable for each column if we want this to mean what it's name suggests. But if we do that, we are simply moving the nullability check up the call stack.

Reply:
> I felt like it added code complexity

Its easy to abstract that complexity. See:

https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/src/groupby/hash/legacy/groupby.cu#L286

> and likely compile time

Certainly, but that's really only a problem if this file is the longest to compile file. Back when I added these 4 different code paths I checked and the impact on compile time was pretty negligible. 

> The template argument can be set to false only when none of the columns are nullable. 

True enough, but that's probably a common scenario. 

> And even then, it only saves a null pointer check.

It saves potentially _two_ null pointer checks per column per call to `aggregate_row`. That can save a lot of overhead given that this function is called in a tight inner loop. 

> I feel the template arguments bool target_has_nulls, bool source_has_nulls are unnecessary. It doesn't indicate if source or target actually has nulls, just that it may.

It means exactly what it says. If `target_has_nulls == true`, then the `target` _table_ contains nulls in one or more of it's columns. Granted, when that get's passed to `update_target_element`, the meaning does change to "target _may_ have nulls". I'd be fine renaming the argument in `update_target_element`. 

> update_target_element has to be able to handle either case anyway.

It does handle both cases, with the ability to optimize for the case when target and/or source columns contain one or more columns with null values. 

> It should ideally select the value of this variable for each column if we want this to mean what it's name suggests.

Ideally, yes, but that's not possible without incurring an additional runtime check per column, per call to `aggregate_row`. 

The optimization is there, so why not use it?

Reply:
Added in https://github.com/rapidsai/cudf/pull/4174/commits/4a04d85f5bd9b5173ffa72146ac78d9d4f529659

>Back when I added these 4 different code paths I checked and the impact on compile time was pretty negligible.

Old no-template method:
```
real    0m49.652s
user    0m46.728s
sys     0m2.920s
```
New template method:
```
real    1m24.104s
user    1m21.017s
sys     0m3.084s
```
It's probably because of thrust. In my experience, using thrust always increases compile time vs making your own kernel.


>Certainly, but that's really only a problem if this file is the longest to compile file.

Tell that to my itty-bitty 12 core. (jk, I don't base my decisions on this)
>That can save a lot of overhead given that this function is called in a tight inner loop.

Groupby sum non nullable column. (potential to save both `nullptr` checks)
No template method:
```
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
Groupby/Basic/100000000/manual_time       24.3 ms         24.3 ms           28
Groupby/Basic/100000000/manual_time       26.0 ms         26.0 ms           28
Groupby/Basic/100000000/manual_time       23.0 ms         23.1 ms           27
```
Template method:
```
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
Groupby/Basic/100000000/manual_time       23.6 ms         23.7 ms           28
Groupby/Basic/100000000/manual_time       27.1 ms         27.1 ms           26
Groupby/Basic/100000000/manual_time       24.7 ms         24.7 ms           31
```

>The optimization is there, so why not use it?

Because it's small fish.

Reply:
Can't argue with benchmarks! The simpler method it is!